### PR TITLE
Disable upgrades on unsupported kernel (fixes #9290)

### LINKED ISSUE #9290: Disable auto upgrades on known unsupported systems
There's a lot of anguish in #9282 and the corresponding forum thread about how we should not allow auto upgrades to versions that don't work -- which, on the face of it, is an entirely reasonable request. What I don't see is anyone volunteering to test every release on a giant lab of systems running unsupported OS versions, so if we're to do anything we need to be a bit more proactive. I suggest the following:

- Our major driver of unsupportedness is the Go compiler and runtime.
- If a platform is supported by Go, and Syncthing runs on it, we consider it "supported" at that point.
- When a platform is no longer supported by Go we consider it "unsupported".
- Once a platform is unsupported, auto upgrades are disabled on that platform.

The last point is something we'll need to implement on our side. Essentially, we'll need to add code to detect the OS version, and when a new Go release comes out we do a last release with the *previous* version plus code that disables auto upgrades on the unsupported platform.

Users with a normal auto upgrade cadence will thus reach this last milestone version and then be stuck there. We should add a startup notice that informs about this fact as well, which could go away when auto upgrades are manually disabled.

Users turning on an unsupported system that's been dormant for a long time might "miss" this release and their Syncthing won't know better than to upgrade to the latest release. If a long time has passed this may fail, like the situation is today. More usually there will be a period of months to years between a platform being "unsupported" and actually ceasing to work, so they'll get one upgrade to a slightly too-new version and get stuck there instead.

Potentially, we could allow an opt out `--allow-unsupported-upgrade` or something for the more intrepid retrocomputing enthusiasts.

*Alternatively*, we could send an OS version header to the upgrade server and let it make the decision. It could then return the latest version supporting that operating system. This would handle systems sleeping over the transition, at the price of sending more data to the upgrade server, and making the server more complicated.

This all seems most relevant for Windows and macOS, which have defined versions with defined end-of-life schedules. There are supportedness issues on the BSDs etc as well, plus with old Linux kernels, but the users will generally know to manage this better than us, I expect.

Discuss.

### PR DESCRIPTION
**Pull request by rasa**
### Purpose

See #9290. ~~This code uses [go/version](https://pkg.go.dev/go/version@go1.22.6?tab=versions) which was first provided in go1.22. Is that acceptable?~~

### Testing

Tested on Linux, macOS Catalina, and Windows.

### Documentation

Do we need to update the docs?

### CODE REVIEW

**Comment by calmh** [comment]
We need to be buildable with Go 1.21 until at least Go 1.23 is released, unfortunately.

**Comment by calmh** [comment]
I didn't check what you're really after but there already is a CompareVersions in the upgrade package, since doing that is part of its core business. It should be mostly applicable.

**Comment by rasa** [comment]
> ... there already is a CompareVersions ...

@calmh Yeah, I switched to that and it actually works better, as we not only need to check for equality, but allow for a minor version mismatch in the CI env.

**Comment by tomasz1986** [comment]
When it comes to Windows, I think one potential problem is that everything starting from the first release of Windows 10 is version `10.0` (including Windows 11 and all Servers), so if Go decides to stop supporting Windows 10 (or older versions of Windows 10), then this will require tweaking to accommodate that (as just `10.0` alone will not be enough).

File: lib/upgrade/upgrade_test.go [hunk 1]
```diff
@@ -179,9 +178,22 @@ func TestCompatibilityJson(t *testing.T) {
 	if compInfo.Runtime != strings.TrimSpace(compInfo.Runtime) {
 		t.Errorf("%s: %q contains spaces", compatibilityJson, compInfo.Runtime)
 	}
-	i := version.Compare(compInfo.Runtime, runtime.Version())
-	if i > 0 {
-		t.Errorf("Got version %s, want %s in %s", compInfo.Runtime, runtime.Version(), compatibilityJson)
+	// If we're running inside a GitHub action, we need to compare the go
+	// version in compatibility.json with both the current runtime version, and
+	// the previous one, as build-syncthing.yaml runs our unit tests with
+	// both versions.
```
Comment (calmh) [inline lib/upgrade/upgrade_test.go:184]: So this brings to light the fact that this check is going to be a pain in the ass for anyone wanting to use a different Go version for whatever reason and not planning to distribute Syncthing or care about the compat file. We should probably have this test run only with a specific `release` tag or something, to make sure _we_ run it but others don't have to.

File: compatibility.json [hunk 2]
```diff
@@ -3,6 +3,8 @@
   "minOSVersion": {
     "darwin": "10.15",
     "linux": "2.6.32",
-    "windows": "10.0"
+    "windows": "10.0",
+    "darwin@go1.23": "11",
+    "linux@go1.24": "3.10"
```
Comment (calmh) [inline compatibility.json:8]: What's this for?

Comment (rasa) [reply compatibility.json:8]: It's just documentation, so the dev who needs to update this file, won't have to do any research, but simply cut-n-paste.

**Comment by rasa** [comment]
> When it comes to Windows, I think one potential problem is that everything starting from the first release of Windows 10 is version `10.0` (including Windows 11 and all Servers), so if Go decides to stop supporting Windows 10 (or older versions of Windows 10), then this will require tweaking to accommodate that (as just `10.0` alone will not be enough).

That's true, but the patch level should allow us to disable Windows 10 or 2016, when and if that ever happens.

File: lib/upgrade/upgrade_test.go (hunk 1 above)
Comment (rasa) [reply lib/upgrade/upgrade_test.go:184]: True. We could add an env var to either opt-in or opt-out. Perhaps opt-out with `STSKIP_COMPATIBILTY_CHECK`?

A dev could always tweak `compatibility.json` to match their env, or even delete the file. Of course, they would need to be careful not to commit their tweak.

Comment (rasa) [reply lib/upgrade/upgrade_test.go:184]: Good idea, I'll add that.

A dev could also tweak `compatibility.json` to match their env, or even delete the file. Of course, they would need to be careful not to commit their tweak.

File: compatibility.json (hunk 2 above)
Comment (calmh) [reply compatibility.json:8]: Mmm let's not do that, it's confusing and how do you know what will be the requirements for Go 1.24? This should be updated at release time, or at least just shortly before.

What we could do to be more flexible is to have a source file with multiple sets of data in it, for example in yaml format:

```
- runtime: go1.21
   requirements:
     darwin: 20
     windows: 10.0
     ...
- runtime: go1.22
   requirements:
     darwin: 21
     windows: 10.1
     ...
```

and then we could at build time pick the set based on runtime.Version, and then it could be a build failure (for everyone) when there aren't any requirements for the current version. Then you could in principle add things for the future as well, though I think we should be careful about that.

Comment (rasa) [reply compatibility.json]: @calmh I implemented your suggestion, so compatibility.json is now generated dynamically from [compatibility.yaml](https://github.com/rasa/syncthing/blob/feature/9290-disable-upgrades/compatibility.yaml) based on the current `runtime.Version()`.

**Comment by calmh** [comment]
I think we can do a refinement here. We can include the compatibility data directly in the upgrade server response, so it's available in the list of releases that the upgrade code looks at. Then it becomes just one more filtering criteria there, and the upgrade code can simply filter out upgrades that are no longer compatible.

**Comment by bt90** [comment]
Delivering the metadata via the server would also have the positive side effect that we can continue to intervene. Otherwise, if we make a mistake, we may already have dozens of installations that no longer update automatically.

**Comment by rasa** [comment]
> ...We can include the compatibility data directly in the upgrade server response ...

@calmh Great idea. Just to confirm, you're saying that instead of shipping `compatibilty.json`, we add its fields to [upgrade.Release](https://github.com/syncthing/syncthing/blob/356c5055ad7596760e553b1dbe524e8af0ee9f3b/lib/upgrade/upgrade_common.go#L22) and return it in `/meta.json`?

**Comment by rasa** [comment]
> We can include the compatibility data directly in the upgrade server response ...

@calmh Somehow the upgrade server will need to know what the `runtime.Version()` of the latest release of syncthing is. We could:

1. The first time the upgrade server sees a new `tag_name` in the GitHub API's response, it downloads the release, untars it, and determines the version of go used to compile the release. It caches this result so it only does this once per tag/release.
2. Immediately restart the upgrade server after a release, making sure it was compiled with the same version of go that was used to compile the release (yuk).
3. Pass the version on the command line (requires an immediate restart).

Which method would you prefer?

The upgrade server would use [debug.ReadBuildInfo](https://github.com/golang/go/blob/1f0c044d60211e435dc58844127544dd3ecb6a41/src/runtime/debug/mod.go#L20) to determine the version of go that was used to compile a binary.

**Comment by calmh** [comment]
The point is to reduce the amount of unnecessary downloads the upgrading Syncthing instance needs to do. I suggest the following:

- The build process creates a compat.json for the runtime used.
- The release process adds compat.json as a release artifact, similar as all the tar.gz files and the signed checksums, etc.
- The upgrade server already reads the list of assets for the release and provides a pared-down view of them. It now just also needs to read compat.json and merge it into its response.
- The upgrading client will get the compat.json for the release together with the list of assets, and can directly compare if this is a release it can upgrade to or not. If not, just skip it same as it would skip a pre-release it's not configured to upgrade to, etc.

**Review by calmh — COMMENTED** [review]
I feel I'm not sure we're envisioning the same solution. Looking the code I'm not sure of the intention. There's stuff to extract the compat.json data from inside the archive, but I think it should be returned by the upgrade server. There's something like it in the upgrade server response, but it does not seem targeted for the release but a general blob of all the compat data?

I think we should:
- Generate a compat.json for the current runtime when building Syncthing
- Have that compat.json be one of the files uploaded as part of the release (this is not in scope for this PR, the release process is separate and I'll deal with it)
- Have the upgrade server look for  (the release specific) compat.json among the assets, when it gets the release metadata from Github, load it, parse it, and add it to the response for that release.
- Have Syncthing look at this when it decides to try to upgrade or not to a given release, before downloading the archive.

**Comment by rasa** [comment]
> I think we should:
> 
> * Generate a compat.json for the current runtime when building Syncthing

@calmh Done.

> * Have that compat.json be one of the files uploaded as part of the release (this is not in scope for this PR, the release process is separate and I'll deal with it)

Gotcha.

> * Have the upgrade server look for  (the release specific) compat.json among the assets, when it gets the release metadata from Github, load it, parse it, and add it to the response for that release.

Gotcha, I will implement this, but it will make the code slightly more complex, but in other ways simpler, too.

tl;dr:

I was trying to avoid the upgrade server having to download anything, as it's just another point of failure, and you and other's liked re @bt90's comment:

> Delivering the metadata via the server would also have the positive side effect that we can continue to intervene. Otherwise, if we make a mistake, we may already have dozens of installations that no longer update automatically.

So the only way to "intervene" is for the upgrade server to always send the contents of the most recent compatibility.yaml, and not the compat.json contained in the release.

But I agree that the upgrade server could download each compat.json for each active release, and using the `runtime` value found, extract the most recent values from compatibility.yaml, and add the values found to the meta.json response. We can cache this info, as it never changes.

> * Have Syncthing look at this when it decides to try to upgrade or not to a given release, before downloading the archive.

Agreed! My apologies for not figuring this out sooner, and thank you for your patience.

**Comment by calmh** [comment]
I think by "intervene" above we meant that we control the upgrade server, so we can change what it does and what it returns if we want, by programming.

**Comment by rasa** [comment]
@calmh OK, I implemented the changes you requested. The upgrade server simply downloads config.json from GitHub and includes its contents in its `meta.json` response.

**Comment by calmh** [comment]
I think I'm failing to communicate here, unfortunately. I do not want a json file in the release archive; I don't want the upgrade server nor Syncthing itself downloading a release archive in order to look for a json file. The compatibility file should be an entirely standalone thing that can be checked before even beginning to download a release archive. I'll take a stab at showing what I mean.

**Comment by calmh** [comment]
This is what I mean #9656

**Comment by rasa** [comment]
@calmh Your solution is much simpler. Sorry I wasn't able to understand the solution you were envisioning, and thanks for reusing my work in your solution. I will test #9656 soon.
